### PR TITLE
Option for showing popup only when key is pressed and configurable delay to open popup.

### PR DESCRIPTION
**Pull request by versusvoid**
Long awaited feature and annoyment reduction.

### CODE REVIEW

**Comment by melink14** [comment]
As a high level comment:
It'd be better if we kept separate features (even if related) in separate pull requests (or perhaps at least separate commits?).  This allows it to be very easy to determine which changes are responsible for which features.

File: manifest.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
 	"manifest_version": 2,
 	
 	"name": "rikaikun",
-	"version": "0.8.91",
+	"version": "0.8.92",
```
Comment (melink14) [inline manifest.json]: Updating this in a PR doesn't really make sense because depending on which features are collected in a release this number will change before publishing.

That being said it's not a big deal because you need some new number for your new options anyway; and it's easy to fix up later before publishing.

File: options.html [hunk 2]
```diff
@@ -30,11 +30,25 @@
...
+							<br />
+							Show popup with delay: 
+							<input type="number" id="popupDelay" name="popupDelay" min="1"> milliseconds
 							</p>
 					</div>
 					<!-- <div id="keycon" class="tabcontent hide"> -->
 					<div id="keycon" class="tabContent">
 					<h1>Keyboard</h1>
+						<input type="checkbox" id="showOnKeyFlag" name="showOnKeyFlag" disabled>
+						Immediately show only when key <select id="showOnKey" name="showOnKey">
```
Comment (melink14) [inline options.html]: This isn't just 'Immediately' right?  Isn't the goal to show definitions *only* when the key is pressed?

File: options.html [hunk 3]
```diff
@@ -30,11 +30,25 @@
...
+							<br />
+							<br />
+							Show popup with delay: 
+							<input type="number" id="popupDelay" name="popupDelay" min="1"> milliseconds
 							</p>
 					</div>
 					<!-- <div id="keycon" class="tabcontent hide"> -->
 					<div id="keycon" class="tabContent">
 					<h1>Keyboard</h1>
+						<input type="checkbox" id="showOnKeyFlag" name="showOnKeyFlag" disabled>
```
Comment (melink14) [inline options.html]: This design seems like it could be simplified.

Instead of a checkbox followed by a select why not use radio buttons where the default value is  "Not used" or similar?

File: options.html [hunk 4]
```diff
@@ -30,11 +30,25 @@
...
+						Immediately show only when key <select id="showOnKey" name="showOnKey"> 
+							<option value="">--</option> 
+							<option value="Alt">Alt</option> 
+							<option value="Ctrl">Ctrl</option> 
+							<option value="Alt+Ctrl">Alt+Ctrl</option> 
+						</select> is pressed.
+
+						<br />
```
Comment (melink14) [inline options.html]: prefer not closing your <br> tags.

File: options.js [hunk 5]
```diff
@@ -82,14 +96,24 @@ function getVals() {
 	localStorage['lineEnding'] = document.optform.lineEnding.value;
 	localStorage['copySeparator'] = document.optform.copySeparator.value;
 	localStorage['maxClipCopyEntries'] = document.optform.maxClipCopyEntries.value;
-
+	try {
+		if (!isFinite(parseInt(document.optform.popupDelay.value))) {
```
Comment (melink14) [inline options.js]: why not save the parsed popupDelay value and then add it to localStorage.  That way you don't need to parseInt twice.

File: rikaicontent.js [hunk 6]
```diff
@@ -236,7 +236,12 @@ var rcxContent = {
 				// go up if necessary
 				if ((y + v + pH) > window.innerHeight) {
 					var t = y - pH - 30;
-					if (t >= 0) y = t;
+					if (t >= 0) {
```
Comment (melink14) [inline rikaicontent.js]: how does this section relate to popup delays and keyboard shortcuts?

File: rikaicontent.js [hunk 7]
```diff
@@ -285,22 +290,22 @@ var rcxContent = {
 			// The reason for this is becasue if there's no selection 
 			// we probably clicked somewhere else and we don't want to 
 			// bounce back.
-		    if(!sel.toString())
```
Comment (melink14) [inline rikaicontent.js]: I appreciate formatting changes but they should be in a separate request.

File: rikaicontent.js [hunk 8]
```diff
@@ -314,12 +319,28 @@ var rcxContent = {
...
+	lastTarget: null,
 
 	onKeyDown: function(ev) { rcxContent._onKeyDown(ev) },
 	_onKeyDown: function(ev) {
-//		this.status("keyCode=" + ev.keyCode + ' charCode=' + ev.charCode + ' detail=' + ev.detail);
-
-		if ((ev.altKey) || (ev.metaKey) || (ev.ctrlKey)) return;
+		//console.log("key=" + ev.key + " keyCode=" + ev.keyCode + ' charCode=' + ev.charCode + ' detail=' + ev.detail);
+ 
+		if (window.rikaichan.config.showOnKey !== "" && (ev.altKey || ev.ctrlKey)) {
```
Comment (melink14) [inline rikaicontent.js]: instead of checking if showOnKey is "", this seems like it would be better to store the keycode of the whichever key was chosen in options right?

File: rikaicontent.js [hunk 9]
```diff
@@ -788,29 +809,46 @@ var rcxContent = {
...
+
+		if ((window.rikaichan.config.showOnKey.includes("Alt") && !ev.altKey) ||
+			 (window.rikaichan.config.showOnKey.includes("Ctrl") && !ev.ctrlKey)) {
+			this.clearHi();
+			this.hidePopup();
+			return;
+		}
+
 		var fake;
+		var tdata = window.rikaichan;	// per-tab data
```
Comment (melink14) [inline rikaicontent.js]: why did you move these vars out?

File: rikaicontent.js [hunk 10]
```diff
@@ -976,9 +1003,13 @@ var rcxContent = {
 			tdata.popX = ev.clientX;
 			tdata.popY = ev.clientY;
 			tdata.timer = setTimeout(
-				function(tdata) {
+				function(tdata, title) {
+					if (!window.rikaichan || title !== window.rikaichan.title) {
+						//console.log("Title changed during delay");
```
Comment (melink14) [inline rikaicontent.js]: remove comment?

File: rikaicontent.js [hunk 11]
```diff
@@ -933,25 +957,28 @@ var rcxContent = {
...
 		}  */
-
 		tdata.prevTarget = ev.target;
 		tdata.prevRangeNode = rp;
 		tdata.prevRangeOfs = ro;
 		tdata.title = null;
 		tdata.uofs = 0;
 		this.uofsNext = 1;
 
+		var delay = window.rikaichan.config.showOnKey === "" ? window.rikaichan.config.popupDelay : 1;
```
Comment (melink14) [inline rikaicontent.js]: why not allow users to set a pop up delay even if they use a showOnKey?

File: rikaicontent.js [hunk 12]
```diff
@@ -933,25 +957,28 @@ var rcxContent = {
...
 			this.forceKanji = ev.shiftKey ? 1 : 0;
 			tdata.popX = ev.clientX;
 			tdata.popY = ev.clientY;
 			tdata.timer = setTimeout(
-				function() {
-					//chrome.extension.sendMessage({"type":"resetDict"});
-					//rcxContent.show(tdata, this.forceKanji ? this.forceKanji : this.defaultDict);
+				function(rangeNode, rangeOffset) {
+					if (!window.rikaichan || rangeNode != window.rikaichan.prevRangeNode || ro != window.rikaichan.prevRangeOfs) {
+						//console.log("Range changed during delay");
```
Comment (melink14) [inline rikaicontent.js]: maybe make this console.log a regular implementation comment for this block?

File: rikaicontent.js [hunk 13]
```diff
@@ -321,19 +321,23 @@ var rcxContent = {
 	keysDown: [],
 	lastPos: { x: null, y: null},
 	lastTarget: null,
+	altGraphKey: false,
 
 	onKeyDown: function(ev) { rcxContent._onKeyDown(ev) },
 	_onKeyDown: function(ev) {
 		//console.log("key=" + ev.key + " keyCode=" + ev.keyCode + ' charCode=' + ev.charCode + ' detail=' + ev.detail);
  
+		if (ev.key === "AltGraph") {
```
Comment (melink14) [inline rikaicontent.js]: Should AltGr be treated as a separate key in the options or is it normal to treat it as identical to alt for these types of things?

File: rikaicontent.js [hunk 14]
```diff
@@ -321,19 +321,23 @@ var rcxContent = {
...
+			this.altGraphKey = true;
+		}
 		if (window.rikaichan.config.showOnKey !== "" && (ev.altKey || ev.ctrlKey)) {
 			if (this.lastTarget !== null) {
 				//console.log(ev);
 				var myEv = {
 					clientX: this.lastPos.x,
 					clientY: this.lastPos.y,
 					target: this.lastTarget,
-					altKey: ev.altKey || ev.key == "Alt",
```
Comment (melink14) [inline rikaicontent.js]: If we want altGraph to be the same as alt here why don't we do something like: ev.altKey || ev.modifierAltGraph

That should work if I'm reading the spec correctly.

I'm not sure we need to add a special variable.

File: options.html (hunk 2 above)
Comment (versusvoid) [inline options.html]: Yes, but without delay, which now defaults to 150ms.

File: rikaicontent.js (hunk 9 above)
Comment (versusvoid) [inline rikaicontent.js]: `tdata` is also used outside `try {}` scope and I have no idea why JS allows such code, but this is definitely wrong.

File: rikaicontent.js (hunk 13 above)
Comment (versusvoid) [inline rikaicontent.js]: Identical. This is chromium's stupidity. On russian layout it generates `Alt` event for left Alt and `AltGraph` for right Alt. Even 'Right Alt+Left' doesn't take you back in history on russian layout, while 'Left Alt+Left' does.

File: rikaicontent.js (hunk 14 above)
Comment (versusvoid) [inline rikaicontent.js]: I'll check if chromium supports it already.

File: rikaicontent.js (hunk 9 above)
Comment (melink14) [inline rikaicontent.js]: Because JS is amazing of course, anything is possible.  (github diffs are hard for me to read so at first I actually didn't see it was outside the try)

This code could really benefit from a re-write since I barely refactored at all when I ported rikaichan.

File: options.html (hunk 2 above)
Comment (melink14) [inline options.html]: I see.  Perhaps we should explicitly mention that it overrides the popup delay setting; at least for me I didn't make the connection.

Also, it seems like no delay is correct for the initial press of the modifier key, but if the user is holding down the modifier key while scanning a sentence, won't they want the same delay as a normal user?

I assume you played with this during implementation and it felt better?

Comment (versusvoid) [inline options.html]: My reasoning was: user has to make _some minimal action_ to show popup. For usual mode I hover mouse and hold it over a word for at least 150ms. But in other mode I'm already doing _some action_ (keeping button pressed), so I expect app response immediately.

But it doesn't matter anymore, I'll make options separate.

File: options.html (hunk 4 above)
Comment (versusvoid) [inline options.html]: Why, by the way? 
xkcd

**Comment by versusvoid** [comment]
@melink14 have fixed everything what can be fixed.

File: options.html (hunk 4 above)
Comment (melink14) [inline options.html]: I guess the core is that it's an extra two characters that aren't needed.

A larger reason is that it's an XHTML idiom that feels out of place in pure
HTML.

I'm sure the requirements of my day job also color my review as well.

It's not really fair to hold everyone to some standard without even a
published style guide. :)

2016年9月9日(金) 12:57 Versus <notifications@github.com>:
